amundsen_application/static/js/utils/navigation-utils.ts

```diff
@@ -0,0 +1,25 @@
+import * as qs from 'simple-query-string';
```

Comment:
What does this mean for other components that -- at least for now -- are calling `window.history`? This is specifically in regards to the `TableDetail` and `ProfilePage` components that use `replaceState` to get rid of logging params. 

It looks like `<Link>` can take an object that has the `pathname` but also a `state`. This would have been helpful to have found earlier: https://github.com/ReactTraining/react-router/blob/master/packages/react-router-dom/docs/api/Link.md#to-object. What if we try using the `state` instead to pass the values we need for logging, and if that works we can get rid of the `window.history` hacks in `TableDetail` and `ProfilePage`.

Reply:
In general I think there should be a level of abstraction between the components window history. At the very least we could write some generic utils to extract logging params for `TableDetail` and `ProfilePage`. 

One limitation of using `state` in `Link` is that it would be limited to in-app navigations. External links would either continue using the url search params or would drop support.

Reply:
Before we consider it logging external page views a limitation, I think it's worth having a discussion about whether or not this is something that is necessary to continue capturing in our action logging.

The approach that we took in `TableDetail` (which was then propagated to `ProfilePage`) was indeed only intended for in-app navigations. Product deemed it valuable to log which options were being clicked on to help measure our search quality, but they were not particularly concerned with capturing all views to the `TableDetail` page in our action logging.

If logging general page views via action logging does become necessary, I think it'd be best to separate from the way we are handling the in-app logging so that we can make the best decisions for custom in-app logging without having to consider page views, and then for page views either leverage google analytics or some custom `logPageView()`. 

Let me know your thoughts, all of these ideas are coming to me in the moment and could be fleshed out more. Do you think it's worth checking in with Product?

Reply:
Especially with some of our newer use cases, I think logging from external URLs has substantial value. 

1) Click rate on Ownership Added/Removed emails
2) Click rate on Details Requested emails
3) Click rate from Lyft Family directory

We could potentially continue using URL params for certain pages and use a utility function to automatically extract and log those page views. I also think we should move away from piggybacking logging params on the primary data request and utilize some explicit function to log events.

Reply:
> I also think we should move away from piggybacking logging params on the primary data request and utilize some explicit function to log events.

I am in support of this in order to give us more control. I think the external URL logging has value, but because -- to my knowledge -- we are not using this information to make product decisions for any features this half I believe it's worth considering doing what is right for logging the in-app clicks on Links without considering external page views, but then also be sure go ahead and do what is correct for external page views. I think we will do ourselves a disservice if we try to maintain one chunk of logic that will work for both cases, as opposed to implementing each case the best way.

Reply:
Also just realized that anything we set up to better log external page views i.e. explicitly calling some `logPageView()` somewhere would also just be how we'd want to log the in-app navigation, potentially just calling `logClick()` with some parameters.

Either way, back to the original concept, are there changes we should be making to `TableDetail` and `ProfilePage` now that we are using this custom history, yet those components still directly call `window.history`?

Reply:
Modifying the code in `TableDetail` and `ProfilePage` can probably be split into a separate task/PR, but what I imagine would be some utility function like


```
// util.ts
export const extractUrlLoggingParams = (keysToExtract) => {
  // extract all keys from url search
  // gather values from sent keys 
  // delete keys from original
  // replace url search with truncated keys
  // return gathered key/value pairs
}

// ProfilePage.tsx
  ... 
  const loggingParams = extractUrlLoggingParams(['source', 'index' ])
  logPageView('/search', loggingParams);
  ...
```

Reply:
1. Separate task/PR is fine tho let's file a ticket. Will things still work without doing that work now? Are there any complications with them still directly calling `window.history` given we are maintaining a custom history?

2. Worth clarifying that what I was getting at was how to do logging without `source` and `index` needing to be in the url and having to replace the url in the first place. Yes it works, but we don't need to do it if we have since introduced better logging support that makes it unnecessary. For example we can call `logSomething(source, index)` when the user clicks on  a `ResourceListItem`, and `logPageView()` if there is a way to know that the user did not just navigate there via the application itself. (perhaps check history). No need to consider this any further in this PR, only getting some ideas out while they come to me. Current concern is to understand the item listed above.

Reply:
1. Filed a ticket. In my testing I haven't encountered any weird interactions with history because of it. I don't completely understand how `createBrowserHistory` works, but I believe it acts as an interface to the native history object, so there is always just one `history` object as the source of truth.

2. I understand what you mean. I wonder if it's worth making a distinction between logging clicks and logging page views. The former would only log in-app clicks, while the latter could log both in-app navigations as well as outside navigations (email, family, etc). If we don't care about external links, we could just log all in-app clicks.